### ISSUE TITLE
Uncart linux compile

### ISSUE THREAD

**Issue by master861** [citra-emu/uncart#16]
I hope this the right place for this frist of hi.
Then i us the least devkitpro i found on the www
My entreypoint to start all things is BootCtr9
Sysnand is 10.0.7-32 i followed this guide https://github.com/Plailect/Guide/wiki/Part-5-%28arm9loaderhax-&-Cakes%29
That is way i dont need emunand.
Now if i compile your uncart version with make it dont work with BootCtr9 or arm9loaderhax can your say me  what for are version this is:The latest release of Uncart for arm9loaderhax.
Or what i need to change that your version works with BootCtr9 or arm9loaderhax.

**Comment by ShadowsFriend** [comment]
If I understand your issue correctly, your self-compiled uncart version doesn't load when started via BootCTR9 on a console running arm9loaderhax. If this is the case I can confirm the issue. While the precompiled uncart version linked in the guide, whose origins seem to remain mysterious, does work, a self-compiled version from this repository's master doesn't. The attached file contains the versions of the used tools and libraries in devkitARM. Furthermore, I'm using python 2.7.10, pycrypto 2.6.1 and make 4.1.

[dkarm-update.txt](https://github.com/citra-emu/uncart/files/188348/dkarm-update.txt)

**Comment by master861** [comment]
Yes that is my  issue your on linux to how the hell your get pycrypto i dont found it for my linux version?
Yes @ShadowsFriend ther exits two versions of uncart how work one on gpatemp and the orther on the tread i postet. 
It comes not from the make version that is used on linux or the python version i compiled Decrypt9WIP by my self and it works.

**Comment by chaoskagami** [comment]
Not to mention whoever made this "mysterious" version needs to release the source to his version, since as of now that version, frankly, is in violation of license.txt. GPL and all that.

**Comment by master861** [comment]
yes That is way i open this tread all things that are on GPL need free open source.
Not like Photoshop ;)
Anyway thx @archshift for changing the name of the tread.

**Comment by gamecaptor** [comment]
"Also, that version trims roms which I find to be undesirable behavior."
@chaoskagam, that's exactly why I was coming here. 

I was using brahma with an emuNAND setup and everything worked fine (dumps where 100%). I then made the move to a9lh with the uncart that is linked in Plailect's guide and now it is trimming them. NOT COOL! 

I can un-trim them with Crappy Tiny Reader, but I would feel better if that version of uncart would leave them intact.

**Comment by chaoskagami** [comment]
I reported an issue on Plailect's guide which was closed, see here: https://github.com/Plailect/Guide/issues/37

In summary: the guy who ported (Gelex on #Cakey) it has to get permission from his company because of his employment contract. Really, he should have gotten permission beforehand, since those contracts are strict without prior permission. I don't feel like taking it up with him, and there's nothing I could do in this legal tape situation, anyways.

I suggest someone else RE his changes and release a version, since in all likelihood, he didn't have permission to commit to a GPL project from his employer. That version should probably be considered an illegal mix of proprietary and GPL code. Expecting a source release at some point is not in our best interest.

**Comment by Plailect** [comment]
https://up1.ca/#gAtGafaAo1u9c9oJtw-CaQ

**Comment by chaoskagami** [comment]
I've built a version based on this patch which only contains changes for a9lh, and not the trimming behavior. https://github.com/chaoskagami/uncart

That said, I guess it isn't really 'trimming' per se - it determines the partition sizes and only dumps up to that. I would wager that is probably the correct behavior, actually.

Part of the patch Plailect posted was d0k3's reboot patch, so I pulled from him rather than apply that bit. Not to mention, the line formatting and spacing was royally screwed so the changes had to be done manually.

Anyways, compiling from my fork should result in a non-trimming a9lh uncart. If anyone here wants me to clean up further and merge to master, tell me what I should do and I'll get it in working order for a merge.

I'm dumping a game to verify it works as-intended, but I don't expect any issues.

**Comment by ShadowsFriend** [comment]
Unfortunately, I can't get your version of uncart to work, @chaoskagami. Neither when booting it via BootCtr9 nor when using the arm9loaderhax.bin directly. Both approaches result in a black screen with backlight on. I'm using AuroraWright's fork of arm9loaderhax (https://github.com/AuroraWright/arm9loaderhax/commit/2b58aff23dcd7b5e93215fb8a25162ddb26e4bc8) with screen-init on an old 3DS. Is there anything I'm missing?

**Comment by chaoskagami** [comment]
You need to run 'make a9lh' as the target, otherwise it compiles the same as upstream. If you previously built with just 'make', run 'make clean' since make won't make any attempt to recompile the incorrect files.

**Comment by ShadowsFriend** [comment]
I actually did use a9lh as target. Just to make sure I cloned your repository in a separate directory and ran make a9lh. This indeed results in the same file that I used for my tests. I used the same toolchain as in https://github.com/citra-emu/uncart/issues/16#issuecomment-200948364.

**Comment by chaoskagami** [comment]
It builds fine on Linux, though admittedly I built devkitARM from source code - the git repo for that matter.

Here's my characteristics for reference:

```
Using built-in specs.
COLLECT_GCC=/arch/ct/dkp/devkitARM/bin/arm-none-eabi-gcc
COLLECT_LTO_WRAPPER=/arch/ct/dkp/devkitARM/libexec/gcc/arm-none-eabi/5.3.0/lto-wrapper
Target: arm-none-eabi
Configured with: ../../gcc-5.3.0/configure --enable-languages=c,c++,objc,obj-c++ --with-gnu-as --with-gnu-ld --with-gcc
--with-march=armv4t --enable-cxx-flags=-ffunction-sections --disable-libstdcxx-verbose --enable-poison-system-directorie
s --enable-interwork --enable-multilib --disable-dependency-tracking --enable-threads --disable-win32-registry --disable
-nls --disable-debug --disable-libmudflap --disable-libssp --disable-libgomp --disable-libstdcxx-pch --target=arm-none-e
abi --with-newlib --with-headers=../../newlib-2.2.0/newlib/libc/include --prefix=/arch/ct/dkp/devkitARM --enable-lto --w
ith-bugurl=http://wiki.devkitpro.org/index.php/Bug_Reports --with-pkgversion='devkitARM release 45'
Thread model: single
gcc version 5.3.0 (devkitARM release 45)
```

I don't know what to suggest here since it seems to be a toolchain issue of some sort. If you just want a working build, I can upload a binary if you'd like, but I'm not sure where to go with debugging from here. I suspect it has nothing to do with uncart.

**Comment by ShadowsFriend** [comment]
I actually don't care that much about a working binary. I want to find and possibly fix the issue :smile:. I'm actually going to try to compile devkitARM from source and see whether that changes anything. Thank you for your help, anyway!

**Comment by chaoskagami** [comment]
Nice to see that you want to fix the issue. Not enough people take that mindset with this kind of stuff. :D

I dunno whether compiling devkitarm will help. Takes a reeeeaaally long time to build, though.

The only other potential difference is delebile a9lh vs AW between our setup. Most people seem to use delebile now, and it does have a few differences.

Anyways, feel free to plug off a message if you need help.

**Comment by ShadowsFriend** [comment]
For testing purposes, I temporarily switched to delebiles fork of arm9loaderhax. Apart from the adrenaline rush associated with updating arm9loaderhax payloads without a hardmod, it unfortunately didn't lead to anything. After devkitARM has finished compiling, I'm going to try some more stuff.

**Comment by chaoskagami** [comment]
Just to make sure; bootctr config is:
```
offset = 0
```
right?

What version of BootCtr9 are you using? If you're not using 0.4.0, you may be hitting https://github.com/hartmannaf/BootCtr9/issues/3.

Otherwise, IDK. The code I have booting up my system leads to everything booting in the same state as arm9loaderhax.bin. Maybe it's BootCtr9.

Also it looks like it was blackscreening because my battery was dead. Hah.

**Comment by ShadowsFriend** [comment]
Yeah, I have the offsets for my arm9payloads set to 0 and I'm using a version of BootCtr9 that was compiled from source a few hours ago. I'm just not sure anymore whether I used the bootloader or the merged branch. I added a few edits to my previous post with some more information in case you missed those.

**Comment by chaoskagami** [comment]
Interesting. The FCRAM fix was, surprisingly, not part of the patch posted by Plailect.

I'm testing a few things - I'll up to another branch when I figure out if it is indeed FCRAM related. You've been very helpful though. :P

**Comment by ShadowsFriend** [comment]
Ha! That seems to have done the trick. Now your version of uncart starts even without the FCRAM workaround in the stage1 payload, i. e. with the latest version of AuroraWright's arm9loaderhax fork. Starting after resets appears to have also become more reliable for it worked three consecutive times after reboots introduced by exiting System Settings.
Thank you very much!

**Comment by chaoskagami** [comment]
Glad to hear it works now. Guess that was the root cause.

For checksums, the No-Intro DB is useful. They host metadata for game dumps, just not the actual games.

I fixed the text issue on my fancy branch. That branch exists mostly because couldn't stand the red-on-white eye-murder anymore... :s Figured I'd add a progressbar while I was at it. Anyways, the fancy branch isn't really relevant to upstream uncart, so any issues relating to that should probably be filed over at my repo instead. But hey, nice to hear you like the UI.

**Comment by ShadowsFriend** [comment]
Apparently, AuroraWright already had an uncart version with an FCRAM fix (https://github.com/AuroraWright/uncart). It seems as if her version contains the whole patchset and not just the minimal changes to make the program work on A9LH, though.

**Comment by chaoskagami** [comment]
Yeah, I noticed too after looking at the network graph.

I don't like trimming though, and a lot of the chqnges of the patch are *really* uneeded. For example, a lot of implicitly unsigned integer constants got a 'u' postfix added unnecessarily. I guess if anybody likes trimming, that'd be the correct fork to use.

On April 23, 2016 3:27:57 AM EDT, ShadowsFriend <notifications@github.com> wrote:
>Apparently, AuroraWright already had an uncart version with an FCRAM
>fix (https://github.com/AuroraWright/uncart). It seems as if her
>version contains the whole patchset and not just the minimal changes to
>make the program work on A9LH, though.
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/citra-emu/uncart/issues/16#issuecomment-213681577

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by gemarcano** [comment]
I'm Gelex. I finally got authorization to release my code publicly (BTW, I wasn't the one distributing the uncart binary being sent around, so don't come yelling at me about GPL compliance. Also, as an aside, you're welcome for using my patches-- Plailect actually got them from me ;-) ). A lot of the "uneeded" changes were made either for consistency, or to make the compiler shut up about warnings related to implicit sign conversion (I did add a lot of extra warnings to the Makefile). I've posted up my fork on my GitHub. I will agree that some things can be cleaned up-- I kinda really just hacked together those changes for trimming to see if I could do it, since it was inspired by what I had seen rom_tool do.

**Comment by chaoskagami** [comment]
To be clear, line endings and spacing were inconsistent in that patch. I couldn't just `patch -Np1` it. It DID allow a9lh compatibility after analyzing what it did. Thanks. :)

I'm not a fan of the trimming behavior but I do admit that it IS useful as an option. Especially for saving on disk space.

How did that binary end up out there? It doesn't really matter since the patch and code are public now, but I'm still curious.

**Comment by gemarcano** [comment]
The line endings and spacing... hm. I recall I tested the patch against my own system and it worked, but w/e. Linux vs. Windows line endings, it's a war with no victors. Now that I'm thinking about it, I do remember some people mentioning this  to me for this particular patch...

What don't you like about the trimming behavior? Seeing that I'm not really trimming so much as only reading all available partitions (including the update partition, which it's debatable whether that one even matters to copy).

The binary got out because I did share it among some people that were tired of dealing with having to break up dumps > 2 GB. Looks like it got shared some more :-). I did end up sharing the patch so people could compile it and have it "spread" the same way the binary did. I was trying to reduce the exposure of my code with the wild until I got permission. TBH, if not for my employer, I would have released all of that code way earlier. I'm just glad to be finally authorized to release my homebrew code in general.

**Comment by chaoskagami** [comment]
It is impossible to apply translation patches to a trimmed rom if they use xdelta3 against a 3ds file, because it will not apply on an incorrect file.

Expanding won't necesarilly fix the checksum, since some information could still be lost. Either way, it isn't ideal when one's goal is archival rather than dumping.

On May 26, 2016 12:07:08 AM EDT, Gabriel Marcano <notifications@github.com> wrote:
>The line endings and spacing... hm. I recall I tested the patch against
>my own system and it worked, but w/e. Linux vs. Windows line endings,
>it's a war with no victors. Now that I'm thinking about it, I do
>remember some people mentioning this  to me for this particular
>patch...
>
>What don't you like about the trimming behavior? Seeing that I'm not
>really trimming so much as only reading all available partitions
>(including the update partition, which it's debatable whether that one
>even matters to copy).
>
>The binary got out because I did share it among some people that were
>tired of dealing with having to break up dumps > 2 GB. Looks like it
>got shared some more :-). I did end up sharing the patch so people
>could compile it and have it "spread" the same way the binary did. I
>was trying to reduce the exposure of my code with the wild until I got
>permission. TBH, if not for my employer, I would have released all of
>that code way earlier. I'm just glad to be finally authorized to
>release my homebrew code in general.
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/citra-emu/uncart/issues/16#issuecomment-221772546

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by gemarcano** [comment]
No information should be lost, unless Nintendo is doing something really, really fucky. All data past the end of the partitions, as far as I can tell, is dumped as 0xFF by uncart. So extending it again should work. For all of my carts this was the case.

**Comment by gemarcano** [comment]
Anyway, if it makes people happy, shouldn't be a problem to provide an option to handle one behavior vs the other.

**Comment by chaoskagami** [comment]
True, it *should* be the case and *is* in every case thus far. That doesn't exclude it from not being the case, as the rom is technically fixed size and could be garbage padded.

I've never seen that, but better safe than sorry. And the NCCH includes the garbage, technically.

**Comment by gemarcano** [comment]
True. The sane solution to satiate both camps would be to offer the option to dump one way or another.

**Comment by chaoskagami** [comment]
I agree 100%. There's a use case for both.

**Comment by TheNain38** [comment]
Can be closed?

**Comment by chaoskagami** [comment]
It looks like this can probably be closed, yeah. #21 seems to resolve the issue.
